netbox/utilities/jobs.py

```diff
@@ -0,0 +1,145 @@
+        """
+        Get all jobs of this `BackgroundJob` related to a specific instance.
+        """
+        jobs = Job.objects.filter(name=cls.name)
+
+        if instance:
+            object_type = ObjectType.objects.get_for_model(instance, for_concrete_model=False)
+            jobs = jobs.filter(
+                object_type=object_type,
+                object_id=instance.pk,
+            )
+
+        return jobs
+
+    @classmethod
+    def enqueue(cls, *args, **kwargs):
+        """
+        Enqueue a new `BackgroundJob`.
+
+        This method is a wrapper of `Job.enqueue()` using `handle()` as function callback. See its documentation for
+        parameters.
+        """
+        return Job.enqueue(cls.handle, name=cls.name, *args, **kwargs)
+
+    @classmethod
+    @advisory_lock(ADVISORY_LOCK_KEYS['job-schedules'])
+    def enqueue_once(cls, instance=None, schedule_at=None, interval=None, *args, **kwargs):
+        """
+        Enqueue a new `BackgroundJob` once, i.e. skip duplicate jobs.
+
+        Like `enqueue()`, this method adds a new `BackgroundJob` to the job queue. However, if there's already a
+        `BackgroundJob` of this class scheduled for `instance`, the existing job will be updated if necessary. This
+        ensures that a particular schedule is only set up once at any given time, i.e. multiple calls to this method are
+        idempotent.
+
+        Note that this does not forbid running additional jobs with the `enqueue()` method, e.g. to schedule an
+        immediate synchronization job in addition to a periodic synchronization schedule.
+
+        For additional parameters see `enqueue()`.
+
+        Args:
+            instance: The NetBox object to which this `BackgroundJob` pertains (optional)
+            schedule_at: Schedule the job to be executed at the passed date and time
+            interval: Recurrence interval (in minutes)
+        """
+        job = cls.get_jobs(instance).filter(status__in=JobStatusChoices.ENQUEUED_STATE_CHOICES).first()
+        if job:
+            # If the job parameters haven't changed, don't schedule a new job and keep the current schedule. Otherwise,
+            # delete the existing job and schedule a new job instead.
+            if (schedule_at and job.scheduled == schedule_at) and (job.interval == interval):
+                return job
+            job.delete()
+
+        return cls.enqueue(instance=instance, schedule_at=schedule_at, interval=interval, *args, **kwargs)
+
+    @classmethod
+    def setup(cls, *args, **kwargs):
+        """
+        Setup a new `BackgroundJob` during plugin initialization.
+
+        This method should be called from the plugins `ready()` function to set up the schedule as early as possible.
+        For interactive setup of schedules (e.g. on user requests), use either `enqueue()` or `enqueue_once()` instead.
+        """
+        connection_created.connect(lambda sender, **signal_kwargs: cls.enqueue_once(*args, **kwargs))
```

Comment:
A minor problem with using the `connection_created' signal is that it's fired on every database connection. This includes every time a job is run by the rqworker. In general this should not have any side effects, but performance will be slightly degraded as the schedules of system background jobs will be checked every time.

To improve performance, I suggest adding an exit-early if `setup()` is not called by a WSGI process, and perhaps only for the first thread. That way, only a subset of NetBox processes will handle the setup of system background jobs. However, I'm not familiar enough with NetBox's WSGI internals to know how to detect these conditions.

Reply:
I may not be clear on the exact functionality here. Would it suffice to just call `enqueue_once()` directly from the AppConfig's `ready()` method, rather than attaching it to a signal?

The `connection_created` signal isn't really tenable for this, as database connections can be created quite frequently depending on the setup, especially when a large number of WSGI workers are in use.

Reply:
True, this is a bit tricky. It feels like there should be a canonical convention for this, but I'm not sure what it would be.

Unfortunately we're on a clock here, as the first v4.1 beta release is due August 5th, and I'd rather not hold up the PR just for this issue. How would you feel about splitting out the auto-run functionality for now? We can merge the remainder of the PR largely as-is (it probably deserves one final review), and revisit the auto-run piece individually. If we arrive at a workable solution, I'd be comfortable slipping that in during the beta period ahead of the final release.

Reply:
That's fine by me. I've removed the `setup()` method and documentation for system jobs. I think we can close this FR without auto-registration and solve this task with #16971.

The code is ready to be used as a system background job, only the auto-registration isn't solved yet. Although not recommended, anyone can already call `enqueue_once()` in the `ready()` method of AppConfig and it *should* work (no guarantee).